File: theMaze_DFS.java
// Time Complexity : O(m*n) where m and n are the dimensions of the maze
// Space Complexity : O(m*n) where m and n are the dimensions of the maze
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class theMaze_DFS {
    int[][] dirs = {{0,1}, {1,0}, {-1,0}, {0, -1}};
    int m, n;
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        if (maze == null || maze.length == 0) return false;
        m = maze.length;
        n = maze[0].length;
        return dfs(maze, start, destination);
    }
    private boolean dfs(int[][] maze, int[] start, int[] destination) {
        int i = start[0], j = start[1];
        // base case
        if (i == destination[0] && j == destination[1]) return true;
        // logic
        maze[i][j] = 2; // mark as visited in place
        for (int[] dir : dirs) {
            int r = i, c = j;
            while (r >= 0 && r < m && c >= 0 && c < n && maze[r][c] != 1) {
                r += dir[0];
                c += dir[1];
            }
            r -= dir[0];
            c -= dir[1];
            if (maze[r][c] != 2 && dfs(maze, new int[] {r, c}, destination)) {
                return true;
            }
        }
        return false;
    }
}

File: TheMaze_DFS.py
# TC: O(M x N) since we will be visiting all the nodes. 
# SC: O(H) where H is the height of the recursive stack.

class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        if not maze or len(maze) == 0: 
            return False
        
        rows = len(maze)
        cols = len(maze[0])
        
        def dfs(i, j):
            if destination[0] == i and destination[1] == j: 
                return True
            
            if maze[i][j] == 2: 
                return False
            
            maze[i][j] = 2
            dirs = [(0,1),(0,-1),(1,0),(-1,0)]
            for x,y in dirs: 
                new_x = i + x
                new_y = j + y
                while new_x >= 0 and new_x < rows and new_y >= 0 and new_y < cols and maze[new_x][new_y] != 1: 
                    new_x += x
                    new_y += y
                
                new_x -= x
                new_y -= y
                if maze[new_x][new_y] != 2: 
                    if dfs(new_x, new_y):
                        return True
            return False
            
        
        return dfs(start[0], start[1])
